From cd0d6f3bb81c2edfd303c5df13ca2ed99e8717d5 Mon Sep 17 00:00:00 2001
Message-Id: <cd0d6f3bb81c2edfd303c5df13ca2ed99e8717d5.1424394676.git.feitong.yi@intel.com>
In-Reply-To: <00f4430a265007e8e2c016cc7926bf530835ea30.1424394676.git.feitong.yi@intel.com>
References: <00f4430a265007e8e2c016cc7926bf530835ea30.1424394676.git.feitong.yi@intel.com>
From: John Harrison <John.C.Harrison@Intel.com>
Date: Tue, 8 Apr 2014 16:57:17 +0100
Subject: [PATCH 29/61] FOR_UPSTREAM [VPG]: drm/i915: Added scheduler
 interrupt handler hook

The scheduler needs to be informed of each batch buffer completion. This is done
via the user interrupt mechanism. The epilogue of each batch buffer submission
updates a sequence number value (seqno) and triggers a user interrupt.

This change hooks the scheduler in to the processing of that interrupt via the
notify_ring() function. The scheduler also has clean up code that needs to be
done outside of the interrupt context, thus notify_ring() now also pokes the
scheduler's work queue.

Change-Id: I4724b3ad7782453a244f84744d54bf14f5b65a38
For: VIZ-1587
For: VIZ-4741
For: GMIN-3638
Signed-off-by: John Harrison <John.C.Harrison@Intel.com>
---
 drivers/gpu/drm/i915/i915_irq.c |    3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/gpu/drm/i915/i915_irq.c b/drivers/gpu/drm/i915/i915_irq.c
index 025ddb1..bded09d 100644
--- a/drivers/gpu/drm/i915/i915_irq.c
+++ b/drivers/gpu/drm/i915/i915_irq.c
@@ -39,6 +39,7 @@
 #include "intel_sync.h"
 #include "intel_drv.h"
 #include "intel_lrc_tdr.h"
+#include "i915_scheduler.h"
 
 static const u32 hpd_ibx[] = {
 	[HPD_CRT] = SDE_CRT_HOTPLUG,
@@ -1271,6 +1272,8 @@ static void notify_ring(struct drm_device *dev,
 	ring->last_irq_seqno = ring->get_seqno(ring, false);
 	trace_i915_gem_request_notify(ring);
 
+	i915_scheduler_handle_irq(ring);
+
 	i915_gem_complete_requests_ring(ring, false);
 
 	if (drm_core_check_feature(dev, DRIVER_MODESET))
-- 
1.7.9.5

